tests/test_views.py

```diff
@@ -1,3 +1,61 @@
-def test_example():
-    print('Tests initialized! Feel free to remove this test and add new ones.')
-    return True
+import pytest
+from pytest_django.asserts import assertTemplateUsed
+from django.urls import reverse
+from asset_dashboard.models import Project
+
+@pytest.mark.django_db
+def test_project_list_view(client, project_list):
+    url = reverse('projects')
+    response = client.get(url)
+    assert response.status_code == 200
+    assert list(project_list) == list(response.context['projects'])
+    assert project_list[0] == response.context['projects'][0]
+
+
+@pytest.mark.django_db
+def test_add_project_view(client):
+    url = reverse('add-edit-projects')
+
+    valid_form_data = {
+        'name': 'trail maintenance project',
+        'description': 'We need to clean the trail and fix some washouts.'
+    }
+
+    successful_response = client.post(url, data=valid_form_data)
+
+    # assert that the new project saved successfully
+    new_project_from_form = Project.objects.filter(name=valid_form_data['name'])[0]
+    assert new_project_from_form.name == valid_form_data['name']
+    assert new_project_from_form.description == valid_form_data['description']
+
+    # user should be redirected to the new project detail page
+    assert successful_response.status_code == 302
+
+    # the new project's detail page should be at this route: /projects/<pk:int>
+    # so make sure the new project PK is in the response's redirect url
+    assert str(new_project_from_form.pk) in successful_response.url
+
+    invalid_form_data = [
```

Comment:
Good to know. I *think* I understand how this would work, but do you know of any DataMade examples doing this? Specifically with testing invalid data.

Reply:
Sure thing. This has a bunch: https://github.com/datamade/large-lots/blob/9b5bcaa48b04e12fbac34901848416a3213c304b/tests/lots_client/test_application.py